File: C++/Patterns/MVC/MVCBenchmark.cpp
#include "AbstractBenchmark.cpp"

namespace oobench {

    class MVCBenchmark : public AbstractBenchmark {
    public:
        MVCBenchmark(int argc, const char** argv) 
                : AbstractBenchmark(argc, argv) { 
        }

        virtual int getMajorNumber() const {
            return 8;
        }

        virtual int getMinorNumber() const {
            return 1;
        }

        virtual void create() = 0;
        virtual void change(unsigned long items) = 0;
        virtual void check(unsigned long items) = 0;
        virtual void remove(unsigned long items) = 0;

        void test(unsigned long items) {
            items = getCountWithDefault(items);
            std::cout << "*** Benchmarking MVC" << std::endl;

            beginAction(1, "creating views", 200, "-O0 due to compilation bug");
            create();
            endAction();

            beginAction(2, "change", items, "-O0 due to compilation bug");
            change(items);
            endAction();

            beginAction(3, "checking views", 200, "-O0 due to compilation bug");
            check(items);
            endAction();

            beginAction(4, "remove", items, "-O0 due to compilation bug");
            remove(items);
            endAction();
        }
    };
}

File: Java/org/oobench/patterns/mvc/MVCBenchmark.java
package org.oobench.patterns.mvc;

import org.oobench.common.*;
import org.oobench.patterns.mvc.common.*;

public abstract class MVCBenchmark extends AbstractBenchmark {

    public int getMajorNumber() {
        return 8;
    }

    public int getMinorNumber() {
        return 1;
    }

    public abstract void create();
    public abstract void change(int items);
    public abstract void check(int items);
    public abstract void remove(int items);

    public void test(int items) {
        items = getCountWithDefault(items);
        System.out.println("*** Benchmarking MVC");

        for (int c = 0; c < 10; c++) {
            if (c == 9) {
                beginAction(1, "creating views", 200);
            }
            create();
            if (c == 9) {
                endAction();
                beginAction(2, "change", items);
            }
            change(items);
            if (c == 9) {
                endAction();
                beginAction(3, "checking views", 200);
            }
            check(items);
            if (c == 9) {
                endAction();
                beginAction(4, "remove", items);
            }
            remove(items);
            if (c == 9) {
                endAction();
            }
        }
    }

    public static void testMVC(Class theClass, int count,
            String[] args) {
        try {
            MVCBenchmark bench = 
                    (MVCBenchmark)theClass.newInstance();
            bench.setArguments(args);
            bench.test(count);
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }
    }
}
